File: src/main/java/me/chriss99/program/TerrainRenderer.java
package me.chriss99.program;

import me.chriss99.TerrainVAO;

import java.util.Collection;

public abstract class TerrainRenderer extends RenderProgram<TerrainVAO> {
    @Override
    public void render(Collection<TerrainVAO> vaos) {
        renderTerrain(vaos);
        renderWater(vaos);
    }

    public abstract void renderTerrain(Collection<TerrainVAO> vao);
    public abstract void renderWater(Collection<TerrainVAO> vao);
}

File: src/main/kotlin/program/TerrainRenderer.kt
package me.chriss99.program

import me.chriss99.render.TerrainVAO

abstract class TerrainRenderer(folder: String = "", vararg fileNames: String) : RenderProgram<TerrainVAO>(folder, * fileNames) {
    override fun render(vaos: Collection<TerrainVAO>) {
        renderTerrain(vaos)
        renderWater(vaos)
    }

    abstract fun renderTerrain(vao: Collection<TerrainVAO>)
    abstract fun renderWater(vao: Collection<TerrainVAO>)
}